src/hotspot/share/runtime/frame.cpp

```diff
@@ -1163,7 +1163,7 @@ void frame::oops_do_internal(OopClosure* f, NMethodClosure* cf,
 #ifndef PRODUCT
   // simulate GC crash here to dump java thread in error report
   if (CrashGCForDumpingJavaThread) {
-    char *t = nullptr;
+    volatile char *t = nullptr; // Use volatile to prevent compiler from optimising away the store
```

Comment:
No, don't do this. We don't need more ways to force a crash. Use
VMError::controlled_crash instead. Note that this will require upgrading that
function to !PRODUCT rather than DEBUG_ONLY. (Don't forget "optimized"
builds.)

OTOH, based on the comment's description of what this is needed for, why not
just `guarantee(!CrashGCForDumpingJavaThread, "")` ?

Reply:
Or do we really need to support this in "optimized" builds?  I'm not sure who uses this.

Reply:
> OTOH, based on the comment's description of what this is needed for, why not just `guarantee(!CrashGCForDumpingJavaThread, "")` ?

Previously I am not quite sure that use `ShouldNotReachHere()` instead of `*t = 'c'` could break the original test intention or not. Use `guarantee(!CrashGCForDumpingJavaThread, "")` make the original test [check](https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/runtime/ErrorHandling/TestDwarf.java#L103) success. So I think use `guarantee(!CrashGCForDumpingJavaThread, "")` will acceptable. The code has been changed. Thanks your advices.